README.md
# probe-rs-rtt

Host side implementation of the RTT (Real-Time Transfer) I/O protocol over probe-rs

## TODO

- Virtual terminal support for channel 0
- Support for filters to limit where to scan for the "control block"
  - Specific memory address (range)
  - Nth block only (if it's duplicated somehow)
  - Symbol address from ELF file?
- Support for using multiple channels at once in the CLI
  - Redirect to file?
  - Redirect to socket?
  - An interactive multi-terminal would be nice but would involve implementing half of "tmux"


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed

### Fixed

## [0.11.0]

### Changed

- Updated to probe-rs 0.11.0

## [0.10.1]

### Fixed

- Fixed a bug where RTT pointers could be torn because of 8bit reads instead of 32bit reads.

## [0.10.0]

### Changed

- Updated to probe-rs 0.10.0

## [0.4.0]

### Added

- Added more logs on all levels.

### Changed

### Fixed

- Fixed a bug where RTT would deadlock.

## [0.3.0]

### Added

- Added a proper warning if no RTT channels are found to be configured.

### Changed

### Fixed

- Fixed some error in the docs.

[Unreleased]: https://github.com/probe-rs/probe-rs/compare/v0.11.0...master
[0.10.1]: https://github.com/probe-rs/probe-rs/compare/v0.10.1...v0.11.0
[0.10.1]: https://github.com/probe-rs/probe-rs/compare/v0.10.0...v0.10.1
[0.10.0]: https://github.com/probe-rs/probe-rs/compare/v0.4.0...v0.10.0
[0.4.0]: https://github.com/probe-rs/probe-rs/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/probe-rs/probe-rs/releases/tag/v0.3.0


rtthost/src/main.rs
use probe_rs::{config::TargetSelector, DebugProbeInfo, Probe};
use probe_rs_rtt::{Channels, Rtt, RttChannel, ScanRegion};
use std::io::prelude::*;
use std::io::{stdin, stdout};
use std::sync::{
    mpsc::{channel, Receiver},
    Arc
};
use std::thread;
use structopt::StructOpt;
use trace_mutex::Mutex;

#[derive(Debug, PartialEq, Eq)]
enum ProbeInfo {
    Number(usize),
    List,
}

impl std::str::FromStr for ProbeInfo {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<ProbeInfo, &'static str> {
        if s == "list" {
            Ok(ProbeInfo::List)
        } else if let Ok(n) = s.parse::<usize>() {
            Ok(ProbeInfo::Number(n))
        } else {
            Err("Invalid probe number.")
        }
    }
}

fn parse_scan_region(mut src: &str) -> Result<ScanRegion, Box<dyn std::error::Error>> {
    src = src.trim();
    if src.is_empty() {
        return Ok(ScanRegion::Ram);
    }

    let parts = src
        .split("..")
        .map(|p| {
            if p.starts_with("0x") || p.starts_with("0X") {
                u32::from_str_radix(&p[2..], 16)
            } else {
                p.parse()
            }
        })
        .collect::<Result<Vec<_>, _>>()?;

    match parts.as_slice() {
        &[addr] => Ok(ScanRegion::Exact(addr)),
        &[start, end] => Ok(ScanRegion::Range(start..end)),
        _ => Err("Invalid range: multiple '..'s".into()),
    }
}

#[derive(Debug, StructOpt)]
#[structopt(
    name = "rtthost",
    about = "Host program for debugging microcontrollers using the RTT (real-time transfer) protocol."
)]
struct Opts {
    #[structopt(
        short,
        long,
        default_value = "0",
        help = "Specify probe number or 'list' to list probes."
    )]
    probe: ProbeInfo,

    #[structopt(
        short,
        long,
        help = "Target chip type. Leave unspecified to auto-detect."
    )]
    chip: Option<String>,

    #[structopt(short, long, help = "List RTT channels and exit.")]
    list: bool,

    #[structopt(
        short,
        long,
        help = "Number of up channel to output. Defaults to 0 if it exists."
    )]
    up: Option<usize>,

    #[structopt(
        short,
        long,
        help = "Number of down channel for keyboard input. Defaults to 0 if it exists."
    )]
    down: Option<usize>,

    #[structopt(
        long,
        default_value="",
        parse(try_from_str=parse_scan_region),
        help = "Memory region to scan for control block. You can specify either an exact starting address '0x1000' or a range such as '0x0000..0x1000'. Both decimal and hex are accepted.")]
    scan_region: ScanRegion,
}

fn main() {
    pretty_env_logger::init();

    std::process::exit(run());
}

fn run() -> i32 {
    let opts = Opts::from_args();

    let probes = Probe::list_all();

    if probes.len() == 0 {
        eprintln!("No debug probes available. Make sure your probe is plugged in, supported and up-to-date.");
        return 1;
    }

    let probe_number = match opts.probe {
        ProbeInfo::List => {
            list_probes(std::io::stdout(), &probes);
            return 0;
        }
        ProbeInfo::Number(i) => i,
    };

    if probe_number >= probes.len() {
        eprintln!("Probe {} does not exist.", probe_number);
        list_probes(std::io::stderr(), &probes);
        return 1;
    }

    let probe = match probes[probe_number].open() {
        Ok(probe) => probe,
        Err(err) => {
            eprintln!("Error opening probe: {}", err);
            return 1;
        }
    };

    let target_selector = opts
        .chip
        .clone()
        .map(|t| TargetSelector::Unspecified(t))
        .unwrap_or(TargetSelector::Auto);

    let session = match probe.attach(target_selector) {
        Ok(session) => session,
        Err(err) => {
            eprintln!("Error creating debug session: {}", err);

            if opts.chip.is_none() {
                if let probe_rs::Error::ChipNotFound(_) = err {
                    eprintln!("Hint: Use '--chip' to specify the target chip type manually");
                }
            }

            return 1;
        }
    };

    eprintln!("Attaching to RTT...");

    let mut rtt = match Rtt::attach_region(Arc::new(Mutex::new(session)), &opts.scan_region) {
        Ok(rtt) => rtt,
        Err(err) => {
            eprintln!("Error attaching to RTT: {}", err);
            return 1;
        }
    };

    if opts.list {
        println!("Up channels:");
        list_channels(rtt.up_channels());

        println!("Down channels:");
        list_channels(rtt.down_channels());

        return 0;
    }

    let up_channel = if let Some(up) = opts.up {
        let chan = rtt.up_channels().take(up);

        if chan.is_none() {
            eprintln!("Error: up channel {} does not exist.", up);
            return 1;
        }

        chan
    } else {
        rtt.up_channels().take(0)
    };

    let down_channel = if let Some(down) = opts.down {
        let chan = rtt.down_channels().take(down);

        if chan.is_none() {
            eprintln!("Error: up channel {} does not exist.", down);
            return 1;
        }

        chan
    } else {
        rtt.down_channels().take(0)
    };

    let stdin = down_channel.as_ref().map(|_| stdin_channel());

    eprintln!("Found control block at 0x{:08x}", rtt.ptr());

    let mut up_buf = [0u8; 1024];
    let mut down_buf = vec![];

    loop {
        if let Some(up_channel) = up_channel.as_ref() {
            let count = match up_channel.read(up_buf.as_mut()) {
                Ok(count) => count,
                Err(err) => {
                    eprintln!("\nError reading from RTT: {}", err);
                    return 1;
                }
            };

            match stdout().write_all(&up_buf[..count]) {
                Ok(_) => {
                    stdout().flush().ok();
                }
                Err(err) => {
                    eprintln!("Error writing to stdout: {}", err);
                    return 1;
                }
            }
        }

        if let (Some(down_channel), Some(stdin)) = (down_channel.as_ref(), &stdin) {
            if let Ok(bytes) = stdin.try_recv() {
                down_buf.extend_from_slice(bytes.as_slice());
            }

            if !down_buf.is_empty() {
                let count = match down_channel.write(down_buf.as_mut()) {
                    Ok(count) => count,
                    Err(err) => {
                        eprintln!("\nError writing to RTT: {}", err);
                        return 1;
                    }
                };

                if count > 0 {
                    down_buf.drain(..count);
                }
            }
        }
    }
}

fn list_probes(mut stream: impl std::io::Write, probes: &Vec<DebugProbeInfo>) {
    writeln!(stream, "Available probes:").unwrap();

    for (i, probe) in probes.iter().enumerate() {
        writeln!(
            stream,
            "  {}: {} {}",
            i,
            probe.identifier,
            probe
                .serial_number
                .as_ref()
                .map(|s| &**s)
                .unwrap_or("(no serial number)")
        )
        .unwrap();
    }
}

fn list_channels(channels: &Channels<impl RttChannel>) {
    if channels.is_empty() {
        println!("  (none)");
        return;
    }

    for chan in channels.iter() {
        println!(
            "  {}: {} (buffer size {})",
            chan.number(),
            chan.name().as_ref().map(|s| &**s).unwrap_or("(no name)"),
            chan.buffer_size(),
        );
    }
}

fn stdin_channel() -> Receiver<Vec<u8>> {
    let (tx, rx) = channel();

    thread::spawn(move || {
        let mut buf = [0u8; 1024];

        loop {
            match stdin().read(&mut buf[..]) {
                Ok(count) => {
                    tx.send(buf[..count].to_vec()).unwrap();
                }
                Err(err) => {
                    eprintln!("Error reading from stdin, input disabled: {}", err);
                    break;
                }
            }
        }
    });

    rx
}


probe-rs-rtt/src/lib.rs
//! Host side implementation of the RTT (Real-Time Transfer) I/O protocol over probe-rs
//!
//! RTT implements input and output to/from a microcontroller using in-memory ring buffers and
//! memory polling. This enables debug logging from the microcontroller with minimal delays and no
//! blocking, making it usable even in real-time applications where e.g. semihosting delays cannot
//! be tolerated.
//!
//! This crate enables you to read and write via RTT channels. It's also used as a building-block
//! for probe-rs debugging tools.
//!
//! ## Example
//!
//! ```no_run
//! use std::sync::{Arc, Mutex};
//! use probe_rs::Probe;
//! use probe_rs_rtt::Rtt;
//!
//! // First obtain a probe-rs session (see probe-rs documentation for details)
//! let probe = Probe::list_all()[0].open()?;
//! let mut session = probe.attach("somechip")?;
//!
//! // Attach to RTT
//! let mut rtt = Rtt::attach(Arc::new(Mutex::new(session)))?;
//!
//! // Read from a channel
//! if let Some(input) = rtt.up_channels().take(0) {
//!     let mut buf = [0u8; 1024];
//!     let count = input.read(&mut buf[..])?;
//!
//!     println!("Read data: {:?}", &buf[..count]);
//! }
//!
//! // Write to a channel
//! if let Some(output) = rtt.down_channels().take(0) {
//!     output.write(b"Hello, computer!\n")?;
//! }
//!
//! # Ok::<(), Box<dyn std::error::Error>>(())
//! ```

use thiserror::Error;

mod channel;
pub use channel::*;

pub mod channels;
pub use channels::Channels;

mod rtt;
pub use rtt::*;

/// Error type for RTT operations.
#[derive(Error, Debug)]
pub enum Error {
    /// RTT control block not found in target memory. Make sure RTT is initialized on the target.
    #[error(
        "RTT control block not found in target memory. Make sure RTT is initialized on the target."
    )]
    ControlBlockNotFound,

    /// Multiple control blocks found in target memory. The data contains the control block addresses (up to 5).
    #[error("Multiple control blocks found in target memory.")]
    MultipleControlBlocksFound(Vec<u32>),

    /// The control block has been corrupted. The data contains a detailed error.
    #[error("Control block corrupted: {0}")]
    ControlBlockCorrupted(String),

    /// Wraps errors propagated up from probe-rs.
    #[error("Error communicating with probe: {0}")]
    Probe(#[from] probe_rs::Error),
}


probe-rs-rtt/src/rtt.rs
use probe_rs::{config::MemoryRegion, MemoryInterface, Session};
use scroll::{Pread, LE};
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::ops::Range;
use std::sync::Arc;
use trace_mutex::Mutex;

use crate::channel::*;
use crate::{Channels, Error};

/// The RTT interface.
///
/// Use [`Rtt::attach`] to attach to a probe-rs `Core` and detect channels.
#[derive(Debug)]
pub struct Rtt {
    ptr: u32,
    up_channels: Channels<UpChannel>,
    down_channels: Channels<DownChannel>,
}

// Rtt must follow this data layout when reading/writing memory in order to be compatible with the
// official RTT implementation.
//
// struct ControlBlock {
//     char id[16]; // Used to find/validate the control block.
//     // Maximum number of up (target to host) channels in following array
//     unsigned int max_up_channels;
//     // Maximum number of down (host to target) channels in following array.
//     unsigned int max_down_channels;
//     RttChannel up_channels[max_up_channels]; // Array of up (target to host) channels.
//     RttChannel down_channels[max_down_channels]; // array of down (host to target) channels.
// }

impl Rtt {
    const RTT_ID: [u8; 16] = *b"SEGGER RTT\0\0\0\0\0\0";

    // Minimum size of the ControlBlock struct in target memory in bytes with empty arrays
    const MIN_SIZE: usize = Self::O_CHANNEL_ARRAYS;

    // Offsets of fields in target memory in bytes
    const O_ID: usize = 0;
    const O_MAX_UP_CHANNELS: usize = 16;
    const O_MAX_DOWN_CHANNELS: usize = 20;
    const O_CHANNEL_ARRAYS: usize = 24;

    fn from(
        session: Arc<Mutex<Session>>,
        memory_map: &[MemoryRegion],
        // Pointer from which to scan
        ptr: u32,
        // Memory contents read in advance, starting from ptr
        mem_in: Option<&[u8]>,
    ) -> Result<Option<Rtt>, Error> {
        let mut mem = match mem_in {
            Some(mem) => Cow::Borrowed(mem),
            None => {
                // If memory wasn't passed in, read the minimum header size
                let mut mem = vec![0u8; Self::MIN_SIZE];
                let mut lock = session.lock().unwrap();
                let mut core = lock.core(0)?;
                core.read_8(ptr, &mut mem)?;
                Cow::Owned(mem)
            }
        };

        // Validate that the control block starts with the ID bytes
        let rtt_id = &mem[Self::O_ID..(Self::O_ID + Self::RTT_ID.len())];
        if rtt_id != Self::RTT_ID {
            log::debug!(
                "Expected control block to start with RTT ID. Got instead: {:?}",
                rtt_id
            );

            return Ok(None);
        }

        let max_up_channels = mem.pread_with::<u32>(Self::O_MAX_UP_CHANNELS, LE).unwrap() as usize;
        let max_down_channels = mem
            .pread_with::<u32>(Self::O_MAX_DOWN_CHANNELS, LE)
            .unwrap() as usize;

        // *Very* conservative sanity check, most people
        if max_up_channels > 255 || max_down_channels > 255 {
            return Err(Error::ControlBlockCorrupted(format!(
                "Nonsensical array sizes at {:08x}: max_up_channels={} max_down_channels={}",
                ptr, max_up_channels, max_down_channels
            )));
        }

        let cb_len = Self::O_CHANNEL_ARRAYS + (max_up_channels + max_down_channels) * Channel::SIZE;

        if let Cow::Owned(mem) = &mut mem {
            // If memory wasn't passed in, read the rest of the control block
            mem.resize(cb_len, 0);
            let mut lock = session.lock().unwrap();
            let mut core = lock.core(0)?;
            core.read_8(
                ptr + Self::MIN_SIZE as u32,
                &mut mem[Self::MIN_SIZE..cb_len],
            )?;
        }

        // Validate that the entire control block fits within the region
        if mem.len() < cb_len {
            log::debug!("Control block doesn't fit in scanned memory region.");
            return Ok(None);
        }

        let mut up_channels = BTreeMap::new();
        let mut down_channels = BTreeMap::new();

        for i in 0..max_up_channels {
            let offset = Self::O_CHANNEL_ARRAYS + i * Channel::SIZE;

            if let Some(chan) =
                Channel::from(&session, i, memory_map, ptr + offset as u32, &mem[offset..])?
            {
                up_channels.insert(i, UpChannel(chan));
            } else {
                log::warn!("Buffer for up channel {} not initialized", i);
            }
        }

        for i in 0..max_down_channels {
            let offset =
                Self::O_CHANNEL_ARRAYS + (max_up_channels * Channel::SIZE) + i * Channel::SIZE;

            if let Some(chan) =
                Channel::from(&session, i, memory_map, ptr + offset as u32, &mem[offset..])?
            {
                down_channels.insert(i, DownChannel(chan));
            } else {
                log::warn!("Buffer for down channel {} not initialized", i);
            }
        }

        Ok(Some(Rtt {
            ptr,
            up_channels: Channels(up_channels),
            down_channels: Channels(down_channels),
        }))
    }

    /// Attempts to detect an RTT control block anywhere in the target RAM and returns an instance
    /// if a valid control block was found.
    ///
    /// `core` can be e.g. an owned `Core` or a shared `Rc<Core>`. The session is only borrowed
    /// temporarily during detection.
    pub fn attach(session: Arc<Mutex<Session>>) -> Result<Rtt, Error> {
        Self::attach_region(session, &Default::default())
    }

    /// Attempts to detect an RTT control block in the specified RAM region(s) and returns an
    /// instance if a valid control block was found.
    ///
    /// `core` can be e.g. an owned `Core` or a shared `Rc<Core>`. The session is only borrowed
    /// temporarily during detection.
    pub fn attach_region(session: Arc<Mutex<Session>>, region: &ScanRegion) -> Result<Rtt, Error> {
        let memory_map: &[MemoryRegion] = &session.lock().unwrap().memory_map().to_vec();

        let ranges: Vec<Range<u32>> = match region {
            ScanRegion::Exact(addr) => {
                log::debug!("Scanning at exact address: 0x{:X}", addr);

                return Rtt::from(session, memory_map, *addr, None)?
                    .ok_or(Error::ControlBlockNotFound);
            }
            ScanRegion::Ram => {
                log::debug!("Scanning RAM");

                memory_map
                    .iter()
                    .filter_map(|r| match r {
                        MemoryRegion::Ram(r) => Some(r.range.clone()),
                        _ => None,
                    })
                    .collect()
            }
            ScanRegion::Range(region) => {
                log::debug!("Scanning region: {:?}", region);

                vec![region.clone()]
            }
        };

        let mut mem: Vec<u8> = Vec::new();
        let mut instances: Vec<Rtt> = Vec::new();

        for range in ranges.iter() {
            if range.len() < Self::MIN_SIZE {
                continue;
            }

            mem.resize(range.len(), 0);
            let mut lock = session.lock().unwrap();
            let mut core = lock.core(0)?;
            core.read_8(range.start, mem.as_mut())?;

            for offset in 0..(mem.len() - Self::MIN_SIZE) {
                if let Some(rtt) = Rtt::from(
                    session.clone(),
                    memory_map,
                    range.start + offset as u32,
                    Some(&mem[offset..]),
                )? {
                    instances.push(rtt);

                    if instances.len() >= 5 {
                        break;
                    }
                }
            }
        }

        if instances.len() == 0 {
            return Err(Error::ControlBlockNotFound);
        }

        if instances.len() > 1 {
            return Err(Error::MultipleControlBlocksFound(
                instances.into_iter().map(|i| i.ptr).collect(),
            ));
        }

        Ok(instances.remove(0))
    }

    /// Returns the memory address of the control block in target memory.
    pub fn ptr(&self) -> u32 {
        self.ptr
    }

    /// Gets the detected up channels.
    pub fn up_channels(&mut self) -> &mut Channels<UpChannel> {
        &mut self.up_channels
    }

    /// Gets the detected down channels.
    pub fn down_channels(&mut self) -> &mut Channels<DownChannel> {
        &mut self.down_channels
    }
}

/// Used to specify which memory regions to scan for the RTT control block.
#[derive(Clone, Debug)]
pub enum ScanRegion {
    /// Scans all RAM regions known to probe-rs. This is the default and should always work, however
    /// if your device has a lot of RAM, scanning all of it is slow.
    Ram,

    /// Limit scanning to these memory addresses in target memory. It is up to the user to ensure
    /// that reading from this range will not read from undefined memory.
    Range(Range<u32>),

    /// Tries to find the control block starting at this exact address. It is up to the user to
    /// ensure that reading the necessary bytes after the pointer will no read from undefined
    /// memory.
    Exact(u32),
}

impl Default for ScanRegion {
    fn default() -> Self {
        ScanRegion::Ram
    }
}


probe-rs-rtt/src/channel.rs
use probe_rs::{config::MemoryRegion, MemoryInterface, Session};
use scroll::{Pread, LE};
use std::cmp::min;
use std::io;
use std::sync::{Arc, Mutex};

use crate::Error;

/// Trait for channel information shared between up and down channels.
pub trait RttChannel {
    /// Returns the number of the channel.
    fn number(&self) -> usize;

    /// Returns the name of the channel or `None` if there is none.
    fn name(&self) -> Option<&str>;

    /// Returns the buffer size in bytes. Note that the usable size is one byte less due to how the
    /// ring buffer is implemented.
    fn buffer_size(&self) -> usize;
}

#[derive(Debug)]
pub(crate) struct Channel {
    session: Arc<Mutex<Session>>,
    number: usize,
    ptr: u32,
    name: Option<String>,
    buffer_ptr: u32,
    size: u32,
}

// Chanels must follow this data layout when reading/writing memory in order to be compatible with
// the official RTT implementation.
//
// struct Channel {
//     const char *name; // Name of channel, pointer to null-terminated string. Optional.
//     char *buffer; // Pointer to buffer data
//     unsigned int size; // Size of data buffer. The actual capacity is one byte less.
//     unsigned int write; // Offset in data buffer of next byte to write.
//     unsigned int read; // Offset in data buffer of next byte to read.
//     // The low 2 bits of flags are used for blocking/non blocking modes, the rest are ignored.
//     unsigned int flags;
// }

impl Channel {
    // Size of the Channel struct in target memory in bytes
    pub(crate) const SIZE: usize = 24;

    // Offsets of fields in target memory in bytes
    const O_NAME: usize = 0;
    const O_BUFFER_PTR: usize = 4;
    const O_SIZE: usize = 8;
    const O_WRITE: usize = 12;
    const O_READ: usize = 16;
    const O_FLAGS: usize = 20;

    pub(crate) fn from(
        session: &Arc<Mutex<Session>>,
        number: usize,
        memory_map: &[MemoryRegion],
        ptr: u32,
        mem: &[u8],
    ) -> Result<Option<Channel>, Error> {
        let buffer_ptr: u32 = mem.pread_with(Self::O_BUFFER_PTR, LE).unwrap();
        if buffer_ptr == 0 {
            // This buffer isn't in use
            return Ok(None);
        }

        let name_ptr: u32 = mem.pread_with(Self::O_NAME, LE).unwrap();

        let name = if name_ptr == 0 {
            None
        } else {
            read_c_string(&mut session.lock().unwrap(), memory_map, name_ptr)?
        };

        Ok(Some(Channel {
            session: Arc::clone(session),
            number,
            ptr,
            name,
            buffer_ptr,
            size: mem.pread_with(Self::O_SIZE, LE).unwrap(),
        }))
    }

    pub fn name(&self) -> Option<&str> {
        self.name.as_ref().map(|s| s.as_ref())
    }

    pub fn buffer_size(&self) -> usize {
        self.size as usize
    }

    fn read_pointers(&self, dir: &'static str) -> Result<(u32, u32), Error> {
        let mut block = [0u32; 2];
        self.session
            .lock()
            .unwrap()
            .core(0)?
            .read_32(self.ptr + Self::O_WRITE as u32, block.as_mut())?;

        let write: u32 = block[0];
        let read: u32 = block[1];

        let validate = |which, value| {
            if value >= self.size {
                Err(Error::ControlBlockCorrupted(format!(
                    "{} pointer is {} while buffer size is {} for {:?} channel {} ({})",
                    which,
                    value,
                    self.size,
                    dir,
                    self.number,
                    self.name().unwrap_or("no name"),
                )))
            } else {
                Ok(())
            }
        };

        validate("write", write)?;
        validate("read", read)?;

        Ok((write, read))
    }
}

/// RTT up (target to host) channel.
#[derive(Debug)]
pub struct UpChannel(pub(crate) Channel);

impl UpChannel {
    /// Returns the number of the channel.
    pub fn number(&self) -> usize {
        self.0.number
    }

    /// Returns the name of the channel or `None` if there is none.
    pub fn name(&self) -> Option<&str> {
        self.0.name()
    }

    /// Returns the buffer size in bytes. Note that the usable size is one byte less due to how the
    /// ring buffer is implemented.
    pub fn buffer_size(&self) -> usize {
        self.0.buffer_size()
    }

    /// Reads the current channel mode from the target and returns its.
    ///
    /// See [`ChannelMode`] for more information on what the modes mean.
    pub fn mode(&self) -> Result<ChannelMode, Error> {
        let mut lock = self.0.session.lock().unwrap();
        let mut core = lock.core(0)?;
        let flags = core.read_word_32(self.0.ptr + Channel::O_FLAGS as u32)?;

        match flags & 0x3 {
            0 => Ok(ChannelMode::NoBlockSkip),
            1 => Ok(ChannelMode::NoBlockTrim),
            2 => Ok(ChannelMode::BlockIfFull),
            _ => Err(Error::ControlBlockCorrupted(String::from(
                "The channel mode flags are invalid",
            ))),
        }
    }

    /// Changes the channel mode on the target to the specified mode.
    ///
    /// See [`ChannelMode`] for more information on what the modes mean.
    pub fn set_mode(&self, mode: ChannelMode) -> Result<(), Error> {
        let mut lock = self.0.session.lock().unwrap();
        let mut core = lock.core(0)?;

        let flags = core.read_word_32(self.0.ptr + Channel::O_FLAGS as u32)?;

        let new_flags = (flags & !3) | (mode as u32);
        core.write_word_32(self.0.ptr + Channel::O_FLAGS as u32, new_flags)?;

        Ok(())
    }

    fn read_core(&self, mut buf: &mut [u8]) -> Result<(u32, usize), Error> {
        let (write, mut read) = self.0.read_pointers("up")?;

        let mut total = 0;

        // Read while buffer contains data and output buffer has space (maximum of two iterations)
        while buf.len() > 0 {
            let count = min(self.readable_contiguous(write, read), buf.len());
            if count == 0 {
                break;
            }

            let mut lock = self.0.session.lock().unwrap();
            let mut core = lock.core(0)?;
            core.read_8(self.0.buffer_ptr + read, &mut buf[..count])?;

            total += count;
            read += count as u32;

            if read >= self.0.size {
                // Wrap around to start
                read = 0;
            }

            buf = &mut buf[count..];
        }

        Ok((read, total))
    }

    /// Reads some bytes from the channel to the specified buffer and returns how many bytes were
    /// read.
    ///
    /// This method will not block waiting for data in the target buffer, and may read less bytes
    /// than would fit in `buf`.
    pub fn read(&self, buf: &mut [u8]) -> Result<usize, Error> {
        let (read, total) = self.read_core(buf)?;

        if total > 0 {
            // Write read pointer back to target if something was read
            let mut lock = self.0.session.lock().unwrap();
            let mut core = lock.core(0)?;
            core.write_word_32(self.0.ptr + Channel::O_READ as u32, read)?;
        }

        Ok(total)
    }

    /// Peeks at the current data in the channel buffer, copies data into the specified buffer and
    /// returns how many bytes were read.
    ///
    /// The difference from [`read`](UpChannel::read) is that this does not discard the data in the
    /// buffer.
    pub fn peek(&self, buf: &mut [u8]) -> Result<usize, Error> {
        Ok(self.read_core(buf)?.1)
    }

    /// Calculates amount of contiguous data available for reading
    fn readable_contiguous(&self, write: u32, read: u32) -> usize {
        (if read > write {
            self.0.size - read
        } else {
            write - read
        }) as usize
    }
}

impl RttChannel for UpChannel {
    /// Returns the number of the channel.
    fn number(&self) -> usize {
        self.0.number
    }

    fn name(&self) -> Option<&str> {
        self.0.name()
    }
    fn buffer_size(&self) -> usize {
        self.0.buffer_size()
    }
}

impl io::Read for UpChannel {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        UpChannel::read(self, buf).map_err(|e| io::Error::new(io::ErrorKind::Other, e))
    }
}

/// RTT down (host to target) channel.
#[derive(Debug)]
pub struct DownChannel(pub(crate) Channel);

impl DownChannel {
    /// Returns the number of the channel.
    pub fn number(&self) -> usize {
        self.0.number
    }

    /// Returns the name of the channel or `None` if there is none.
    pub fn name(&self) -> Option<&str> {
        self.0.name()
    }

    /// Returns the buffer size in bytes. Note that the usable size is one byte less due to how the
    /// ring buffer is implemented.
    pub fn buffer_size(&self) -> usize {
        self.0.buffer_size()
    }

    /// Writes some bytes into the channel buffer and returns the number of bytes written.
    ///
    /// This method will not block waiting for space to become available in the channel buffer, and
    /// may not write all of `buf`.
    pub fn write(&self, mut buf: &[u8]) -> Result<usize, Error> {
        let (mut write, read) = self.0.read_pointers("down")?;

        if self.writable_contiguous(write, read) == 0 {
            // Buffer is full - do nothing.
            return Ok(0);
        }

        let mut total = 0;

        // Write while buffer has space for data and output contains data (maximum of two iterations)
        while buf.len() > 0 {
            let count = min(self.writable_contiguous(write, read), buf.len());
            if count == 0 {
                break;
            }

            let mut lock = self.0.session.lock().unwrap();
            let mut core = lock.core(0)?;
            core.write_8(self.0.buffer_ptr + write, &buf[..count])?;

            total += count;
            write += count as u32;

            if write >= self.0.size {
                // Wrap around to start
                write = 0;
            }

            buf = &buf[count..];
        }

        // Write write pointer back to target

        let mut lock = self.0.session.lock().unwrap();
        let mut core = lock.core(0)?;
        core.write_word_32(self.0.ptr + Channel::O_WRITE as u32, write)?;

        Ok(total)
    }

    /// Calculates amount of contiguous space available for writing
    fn writable_contiguous(&self, write: u32, read: u32) -> usize {
        (if read > write {
            read - write - 1
        } else if read == 0 {
            self.0.size - write - 1
        } else {
            self.0.size - write
        }) as usize
    }
}

impl RttChannel for DownChannel {
    /// Returns the number of the channel.
    fn number(&self) -> usize {
        self.0.number
    }

    fn name(&self) -> Option<&str> {
        self.0.name()
    }
    fn buffer_size(&self) -> usize {
        self.0.buffer_size()
    }
}

impl io::Write for DownChannel {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        DownChannel::write(self, buf).map_err(|e| io::Error::new(io::ErrorKind::Other, e))
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

/// Reads a null-terminated string from target memory. Lossy UTF-8 decoding is used.
fn read_c_string(
    session: &mut Session,
    memory_map: &[MemoryRegion],
    ptr: u32,
) -> Result<Option<String>, Error> {
    // Find out which memory range contains the pointer
    let range = memory_map
        .iter()
        .filter_map(|r| match r {
            MemoryRegion::Flash(r) => Some(&r.range),
            MemoryRegion::Ram(r) => Some(&r.range),
            _ => None,
        })
        .find(|r| r.contains(&ptr));

    // If the pointer is not within any valid range, return None.
    let range = match range {
        Some(r) => r,
        None => return Ok(None),
    };

    // Read up to 128 bytes not going past the end of the region
    let mut bytes = vec![0u8; min(128, (range.end - ptr) as usize)];
    session.core(0)?.read_8(ptr, bytes.as_mut())?;

    // If the bytes read contain a null, return the preceding part as a string, otherwise None.
    Ok(bytes
        .iter()
        .position(|&b| b == 0)
        .map(|p| String::from_utf8_lossy(&bytes[..p]).into_owned()))
}

/// Specifies what to do when a channel doesn't have enough buffer space for a complete write on the
/// target side.
#[derive(Eq, PartialEq, Debug)]
#[repr(u32)]
pub enum ChannelMode {
    /// Skip writing the data completely if it doesn't fit in its entirety.
    NoBlockSkip = 0,

    /// Write as much as possible of the data and ignore the rest.
    NoBlockTrim = 1,

    /// Block (spin) if the buffer is full. Note that if the application writes within a critical
    /// section, using this mode can cause the application to freeze if the buffer becomes full and
    /// is not read by the host.
    BlockIfFull = 2,
}
